### PR TITLE
Build-test documentation in the CI and fix/improve broken links

### PR DESCRIPTION
**Pull request by MarijnS95**
Some intradoc links were broken because their crates are only available as `dev-dependencies`, while others unnecessarily had a forced link to `docs.rs` which could also be an intradoc link (in turn pointing to locally built documentation if not building on docs.rs).

To make sure these links don't break again over time, build-test the documentation in the CI while disallowing any warnings which includes broken links.

### CODE REVIEW

**Review by loewenheim — APPROVED** [review]

File: .github/workflows/ci.yml
Comment (flub) [inline .github/workflows/ci.yml:168]: Does this make sense for docs?  E.g. in symbolicator we check the doc comments simply on `runs-on: ubuntu-latest` and with stable rust

File: .github/workflows/ci.yml
Comment (flub) [inline .github/workflows/ci.yml:182]: I think you can add `components: rust-docs` here to make this a little faster.

File: .github/workflows/ci.yml
Comment (flub) [reply .github/workflows/ci.yml:192]: Currently there seems to be no errors when adding `--document-private-items` so I would be fine with adding it for now.  If it becomes a pain we can always remove it again.

**Review by flub — COMMENTED** [review]
thanks for adding this!

File: .github/workflows/ci.yml
Comment (MarijnS95) [reply .github/workflows/ci.yml:168]: For these specific tests we can do it without, as long as `--cfg doc_cfg` isn't replicated from `docs.rs` setup.

File: .github/workflows/ci.yml
Comment (MarijnS95) [reply .github/workflows/ci.yml:182]: Isn't that additional, on top of what is already installed by `profile: minimal`?

(I'm surprised that this works, since `rust-docs` is only included in https://rust-lang.github.io/rustup/concepts/profiles.html - perhaps it's pulling it from GH's runner/)

Comment (MarijnS95) [reply .github/workflows/ci.yml:182]: EDIT: Perhaps it doesn't work. The CI in this PR seems to run the workflow from `master`, so _excluding_ my job (something to do with possibly exposing secrets without maintainer approval, I suppose), and I wouldn't have caught this as the `on:` branch filter doesn't allow my branchname.

File: .github/workflows/ci.yml
Comment (flub) [reply .github/workflows/ci.yml:168]: I'm not sure what you're saying here, you are adding `--cfg doc_cfg` below?  Is it preferable to check the docs on all platforms and with nightly?

I'm just worried we're blowing up the amount of CI resources for no real benefit with this matrix.

Comment (MarijnS95) [reply .github/workflows/ci.yml:168]: 1. `nightly`: This is needed to enable `--cfg doc_cfg` for `#![cfg_attr(doc_cfg, feature(doc_cfg))]`. It doesn't seem worth it, since `rustdoc` doesn't validate `doc(cfg())` features to actually exist :(;
2. Having 3 `os`'es in a matrix is not needed, there are only very few os `cfg`s that have no effect on documentation. I'll remove them.

Comment (MarijnS95) [reply .github/workflows/ci.yml]: Sorry for all the noise (repeated pushes) in all PRs, great lack of focus for the simplest of things :weary:

**Review by flub — APPROVED** [review]

File: .github/workflows/ci.yml
Comment (flub) [reply .github/workflows/ci.yml]: Oh, I forgot that `--cfg doc_cfg` needs nightly.  thanks for reminding me :)

**Pull request merged**
